Criado rota de login pela API

TTOExtreme opened the pull request:
-- Adicionado o router por proxy para o modo DEV (criado pelo GCastilho)
-- Criado rota de login  api.<Dominio>/v1/login
    - Dentro da rota é passado o SessionId para o cookie dentro do subdominio API

Commit:
Criado rota de login pela API

-- Adicionado o router por proxy para o modo DEV (criado pelo GCastilho)
-- Criado rota de login  api.<Dominio>/v1/login
    - Dentro da rota é passado o SessionId para o cookie dentro do subdominio API

Commit:
Options somente no SubDominio API

GCastilho commented on src/server/index.ts:
```diff
@@ -38,7 +28,18 @@ app.use((req, res, next) => {
 	case('api'):
 		/** Request to the api subdomain */
-		return api(req, res, next)
+		if(req.method == 'OPTIONS'){
+			/**
+			 * necessario para passar pelo preflight check
+			 * o preflight checa se a operação pode ser feita (POST, GET, DELETE ...)
+			 */
+			res.header('Access-Control-Allow-Origin', 'http://localhost:3000')
+			res.header('Access-Control-Allow-Credentials','true')
+			res.header('Access-Control-Allow-Headers', 'Content-Type')
+			return res.sendStatus(200)
+		}else{
+			return api(req, res, next)
+		}
```
Coloca isso dentro do index da v1 ou da v1/login usando req.options

GCastilho replied:
Pq esse arquivo é só um roteador, não é pra ficar nada além disso nele

GCastilho commented on src/server/api/v1/login.ts:
```diff
@@ -0,0 +1,86 @@
+const router = express.Router()
+
+// Parsers
+router.use(bodyParser())
+router.use(cookieParser())
+router.use(express.json())
```
espress.json() é um bodyparser deprected, vai ser removido, usa bodyparser.json()
'bodyparser()' sem nada tbm não é um middleware válido, precisa de um extended

Commit:
Mudança do options para dentro da rota

-- Alterado o local do options para dentro do /v1/login

GCastilho committed:
Corrigido problemas com o CORS

-- Corrigido problema com o cookie e com o hanlder do preflight que estavam fazendo o cookie não estar autorizado para ser usado no domínio da api

GCastilho committed:
Movido autenticação para /v1/user/authentication

-- Motivo handler de autenticação para /v1/user/authentication

GCastilho committed:
Adicionado handler de desautenticação

-- Refatorado o hanlder de autenticação para usar async/await e não cadeia de promessas
-- Adicionado um handler de desautenticação em /v1/user/authentication (ao usar DELETE)
-- Adicionado notas para implementação dos testes novos (necessário refatorar os testes)
-- Os métodos antigos de autenticação não foram removidos pois ainda são usando nos testes de update de senha

GCastilho committed:
Corrigido comentário desalinhado

GCastilho committed:
Refatorado handler do preflight

-- Movido middleware do CORS para a api/v1
-- Refatorado hanlder do preflight q agora usa o middleware da API para setar os headers

GCastilho reviewed (approved):
Precisa fazer os testes pro novo autenticador mas não dá pra fazer isso enquanto os testes não forem refatorados (pq eles tão uma zona) então eu adicionei uma nota pra lembrar q tem q fazer eles e segue a vida
